### ISSUE TITLE
SubscribeToPushNotifications

### ISSUE THREAD

**Issue by synic** [ecederstrand/exchangelib#145]
https://msdn.microsoft.com/en-us/library/office/aa566188(v=exchg.150).aspx#Anchor_3

Love the library!  Man I wish you had posted this a few years ago, could have saved me a lot of trouble!

I was thinking about adding the `SubscribeToPushNotifications` operation somehow, but not totally sure where to start.

From reading the code, it looks like I'd add a "service" to `services.py`, called `SubscriptionRequest`, and then several items to `properties.py` - does that sound right?

Where do you think a utility method like this would live in relation to the `Account`?  `Account.subscribe` or `Account.protocol.subscribe` or something like that?

**Comment by ecederstrand** [comment]
Yes, you'd add a new `SubscriptionRequest` service. The service operates on the account level, taking a list of folders, so `Account` would be the natural place to add a utility method (`Account.protocol` is a shortcut to services that are not account-specific, like server timezone info).

It would probably be nice to also have a `Folder.subscribe()` method that is just a thin wrapper for `Account.subscribe()`.

Implementing the other half of subscriptions, i.e. consuming the push notifications, would be nice, but that can wait.

**Comment by synic** [comment]
Awesome.  I'll mess around with it, hopefully eventually have a PR for you.

**Comment by emasunag** [comment]
Hey, so i wanted to subscribe to new emails and calendar events using exchange lib. I didn't want a post a new issue since this is here, but I am still confused on how to go about this.

**Comment by ecederstrand** [comment]
Basically, it's noit implemented yet. There is some work going on with notifications support in https://github.com/ecederstrand/exchangelib/pull/355 but it's not ready for merge yet.

**Comment by cs-shivaji-kolape** [comment]
Hey, @ecederstrand 
I have written a script for the Exchange Notification service using the EWS API and it is working fine, but it has been failing frequently in recent days and returns an error response below. This is tried in both access types (delegated and impersonation).

```
<?xml version="1.0" encoding="utf-8"?><s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Header><h:ServerVersionInfo MajorVersion="15" MinorVersion="20" MajorBuildNumber="3763" MinorBuildNumber="14" Version="V2018_01_08" xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"/></s:Header><s:Body><m:SubscribeResponse xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><m:ResponseMessages><m:SubscribeResponseMessage ResponseClass="Error"><m:MessageText>You have exceeded the available subscriptions for your account.  Remove unnecessary subscriptions and try your request again.</m:MessageText><m:ResponseCode>ErrorExceededSubscriptionCount</m:ResponseCode><m:DescriptiveLinkKey>0</m:DescriptiveLinkKey><m:MessageXml><t:Value Name="PolicyLimit">20</t:Value></m:MessageXml></m:SubscribeResponseMessage></m:ResponseMessages></m:SubscribeResponse></s:Body></s:Envelope>
```

any help would be great.

**Comment by ecederstrand** [comment]
Googling a bit, it seems that EWS doesn't allow you to list the current subscriptions, and you can't unsubscribe unless you have the subscription ID. Subscribers need to respond to heartbeats for the subscription to stay alive, so it seems you have active subscribers somewhere. Using impersonation is supposed to overcome this limitation.

**Comment by ecederstrand** [comment]
Push and stream notification support was added in a series of commits ending in a729eac
